### PR TITLE
Fix: Final URL to managed access manifests is broken (#3922, #3328)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Connected issue: #3922 

### Author

- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                          <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

## Author (hotfixes)

- [x] Added `h` tag to commit title and PR targets `prod`           <sub>or this PR does not include a temporary hotfix</sub>
- [x] Added `H` tag to commit title                                 <sub>or this PR does not include a permanent hotfix</sub>
- [x] Added `hotfix` label to PR                                    <sub>or this PR does not include a hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [x] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                         <sub>or PR description does not contain an announcement</sub>
- [ ] Shortened the PR chain                                        <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate               <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved connected issue to `prod` or `Merged prod`              <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                          <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                             <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

File: src/azul/service/manifest_service.py [hunk 1]
```diff
@@ -450,7 +450,9 @@ def get_cached_manifest_with_object_key(self,
                                                  authentication)
         current_source_key = generator.compute_source_key()
         manifest_key, source_key, extension = object_key.rsplit('/', 1)[-1].split('.')
-        if source_key != current_source_key:
+        # FIXME: Add support for long-lived API tokens
+        #        https://github.com/DataBiosphere/azul/issues/3328
```
Comment (achave11) [inline src/azul/service/manifest_service.py:454]: This comment should be added to an additional commit titled "Add FIXME (#3328)" (per our convention).

**Review by achave11 — COMMENTED** [review]
Another reason for adding a second commit is that the present commit doesn't resolve (#3328) AFAICT, so I am unsure it is accurate to reference it in this form in the commit or the PR tittle.

**Review by achave11 — APPROVED** [review]
LGTM!

**Review by hannes-ucsc — COMMENTED** [review]
I blocked the connected ticket on #2956 for which there is another PR in the works (#3925). Please hold this warm until all blockers are in `dev`.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I think this leaves the source code in a bit of a confusing state. With this surgical approach, the service still goes through all the motions of actually listing the accessible sources in the controller and populating the filter object with the result even though the wrong credentials are used to request the listing.

Maybe this makes it more clear

```diff
Index: src/azul/service/source_controller.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/source_controller.py b/src/azul/service/source_controller.py
--- a/src/azul/service/source_controller.py	(revision 7e6b6143b13fc42cd9ed60dd76d70fd52ad68950)
+++ b/src/azul/service/source_controller.py	(date 1648084946401)
@@ -54,7 +54,7 @@
 
     def get_filters(self,
                     catalog: CatalogName,
-                    authentication: Authentication,
+                    authentication: Optional[Authentication],
                     filters: Optional[str] = None
                     ) -> MutableFilters:
         return MutableFilters(explicit=self._parse_filters(filters),
Index: src/azul/service/manifest_controller.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/manifest_controller.py b/src/azul/service/manifest_controller.py
--- a/src/azul/service/manifest_controller.py	(revision 7e6b6143b13fc42cd9ed60dd76d70fd52ad68950)
+++ b/src/azul/service/manifest_controller.py	(date 1648084888573)
@@ -107,10 +107,10 @@
         token = query_params.get('token')
         if token is None:
             format_ = ManifestFormat(query_params['format'])
-            filters = self.get_filters(catalog, authentication, query_params.get('filters'))
             try:
                 object_key = query_params['objectKey']
             except KeyError:
+                filters = self.get_filters(catalog, authentication, query_params.get('filters'))
                 object_key, manifest = self.service.get_cached_manifest(format_=format_,
                                                                         catalog=catalog,
                                                                         filters=filters,
@@ -128,6 +128,10 @@
                     }
                     token = self.async_service.start_generation(state)
             else:
+                # FIXME: Add support for long-lived API tokens
+                #        https://github.com/DataBiosphere/azul/issues/3328
+                assert authentication is None, authentication
+                filters = self.get_filters(catalog, authentication, query_params.get('filters'))
                 try:
                     manifest = self.service.get_cached_manifest_with_object_key(
                         format_=format_,
```

But check the `Optional[Authentication]` change and look for more places where that is wrong.

Squash the FIXME commit and the main commit. The `Optional[Authentication]` changes should be a separate commit.

**Comment by hannes-ucsc** [comment]
I'm assuming the intention is for me to be assigned.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
```diff
Index: src/azul/service/manifest_service.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/manifest_service.py b/src/azul/service/manifest_service.py
--- a/src/azul/service/manifest_service.py	(revision 4885cf2f20167ff2039f547f8b6a9221861331fe)
+++ b/src/azul/service/manifest_service.py	(date 1649120207370)
@@ -451,12 +451,15 @@
                                                  catalog,
                                                  filters,
                                                  authentication)
-        current_source_key = generator.compute_source_key()
-        manifest_key, source_key, extension = object_key.rsplit('/', 1)[-1].split('.')
-        # FIXME: Add support for long-lived API tokens
-        #        https://github.com/DataBiosphere/azul/issues/3328
-        if False and source_key != current_source_key:
-            raise CachedManifestSourcesChanged
+        # FIXME: Add support for long-lived API tokens
+        #        https://github.com/DataBiosphere/azul/issues/3328
+        if False:
+            current_source_key = generator.compute_source_key()
+            # FIXME: Consolidate parsing of manifest object key
+            #        REVIEW: insert link to ticket
+            manifest_key, source_key, extension = object_key.rsplit('/', 1)[-1].split('.')
+            if source_key != current_source_key:
+                raise CachedManifestSourcesChanged
         file_name = self._get_cached_manifest_file_name(generator, object_key)
         if file_name is None:
             raise CachedManifestNotFound
@@ -918,6 +921,8 @@
             timestamp = datetime.now().strftime("%Y-%m-%d %H.%M")
             file_name = f'{file_name_prefix} {timestamp}.{self.file_name_extension}'
         else:
+            # FIXME: Consolidate parsing of manifest object key
+            #        REVIEW: insert link to ticket
             file_name = 'hca-manifest-' + object_key.rsplit('/', )[-1]
         return file_name
 
Index: src/azul/service/manifest_controller.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/manifest_controller.py b/src/azul/service/manifest_controller.py
--- a/src/azul/service/manifest_controller.py	(revision 4885cf2f20167ff2039f547f8b6a9221861331fe)
+++ b/src/azul/service/manifest_controller.py	(date 1649120524595)
@@ -133,8 +133,7 @@
                 if authentication is None:
                     filters = self.get_filters(catalog, authentication, query_params.get('filters'))
                 else:
-                    raise BadRequestError(f'Authentication is not accepted when using the `objectKey` parameter. '
-                                          f'Recieved authentication {authentication.as_http_header()}')
+                    raise BadRequestError("Must omit authentication when passing 'objectKey'")
                 try:
                     manifest = self.service.get_cached_manifest_with_object_key(
                         format_=format_,
```

**Review by hannes-ucsc — APPROVED** [review]
~I think the two commits~

**Pull request merged**